Added new services to platform kodi

alexmogavero opened the pull request:
## Description:
Some services have been added to component media_player and have been implemented in platform kodi.
These additional services enable search and play through the music library, together with some additional controls. For instance they can be used to couple kodi with alexa.
Here there is the list of additional services:

- play_song
- add_song_to_playlist
- add_album_to_playlist
- set_shuffle
- unset_shuffle
- add_all_albums_to_playlist


**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#<home-assistant.github.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

alexmogavero committed:
added new service

alexmogavero committed:
fixed basic test in kodi platform

alexmogavero committed:
Added new method async_get_albums

alexmogavero committed:
Added new methods in module kodi

alexmogavero committed:
added method find_song in kodi module

alexmogavero committed:
method add_song_to_playlist made
async

alexmogavero committed:
Added media type to method async_play_media

alexmogavero committed:
added methods async_clear_playlist
and play_song

alexmogavero committed:
methods play_song and find_song
made async

alexmogavero committed:
added new service play_song

alexmogavero committed:
Merge branch 'dev' into kodi

alexmogavero committed:
Improved kodi._find
now it find for whole words only

alexmogavero committed:
added possibility to specify artist in
kodi.async_find_artist

alexmogavero committed:
added kodi.async_find_album

alexmogavero committed:
added new optional input to play_song service

alexmogavero committed:
In async_play_song added handling of no song found

alexmogavero committed:
default artist value changed to ''

alexmogavero committed:
async_add_song_to_playlist now can also
search for musinc

alexmogavero committed:
added service add_song_to_playlist

alexmogavero committed:
Added new service add_album_to_playlist

alexmogavero committed:
added services to switch shuffle mode

alexmogavero committed:
added service add_all_albums_to_playlist

alexmogavero committed:
handled error in async_unset_shuffle
and async_set_shuffle

alexmogavero committed:
Added abstract methods to media_player

alexmogavero committed:
Merge branch 'dev' into kodi

# Conflicts:
#	homeassistant/components/media_player/kodi.py

alexmogavero committed:
_server substituted with server property

alexmogavero committed:
style made consistent with requirements

alexmogavero committed:
Fixed issue with pylint

MartinHjelmare commented on homeassistant/components/media_player/__init__.py:
```diff
@@ -133,6 +145,27 @@
+MEDIA_PLAYER_ADD_SONG_SCHEMA = MEDIA_PLAYER_SCHEMA.extend({
+    vol.Optional(ATTR_MEDIA_SONG_ID): cv.string,
+    vol.Optional(ATTR_MEDIA_SONG_NAME): cv.string,
+    vol.Optional(ATTR_MEDIA_ARTIST_NAME): cv.string,
```
Shouldn't at least one of these be required?

alexmogavero replied:
No this should work also without anything, even though the output would be simply the first song in the library.

MartinHjelmare commented on homeassistant/components/media_player/__init__.py:
```diff
@@ -133,6 +145,27 @@
+MEDIA_PLAYER_ADD_ALBUM_SCHEMA = MEDIA_PLAYER_SCHEMA.extend({
+    vol.Optional(ATTR_MEDIA_ALBUM_ID): cv.string,
+    vol.Optional(ATTR_MEDIA_ALBUM_NAME): cv.string,
+    vol.Optional(ATTR_MEDIA_ARTIST_NAME): cv.string,
```
See above.

alexmogavero replied:
Like above.

MartinHjelmare commented on homeassistant/components/media_player/__init__.py:
```diff
@@ -133,6 +145,27 @@
+})
+
+MEDIA_PLAYER_ADD_ALL_ALBUMS_SCHEMA = MEDIA_PLAYER_SCHEMA.extend({
+    vol.Optional(ATTR_MEDIA_ARTIST_NAME): cv.string,
```
You have this as a positional argument in the corresponding method.

alexmogavero replied:
You are right. I will amend it, thank you!

MartinHjelmare reviewed (commented):
For each service added you should add a function in the component root that calls the corresponding service.

alexmogavero commented:
@MartinHjelmare no problem to add a function for each service. However can I ask why is this needed? It seems to work perfectly also without.

MartinHjelmare commented:
It's to be able to import the media_player module in another module and call the services, easily.

There should be a motivation to reuse services in multiple platforms when adding services to the base component. Will these added services work with other platforms? I don't use media players in HA yet, so I don't have a specific opinion in this case. The alternative is to add the services to the platform that needs them.

Before putting more work into this direction you might want to get a second opinion from another HA developer.

pvizeli commented:
I'm not exact sure if that okay -> @balloob. I think it was plan to extend base media_player with snapshot/restore and group functions. This is to platform specific. So disrate this to add that to base.

For name schema and other things, look to `sonos.py` how that should look out.

alexmogavero commented:
I have added the services in order to be used with Alexa.
So I needed the services to be visible from the `configuration.yaml` file.
Then I added the methods in the base class so to raise an error if anyone tries to call the service from another platform.
@pvizeli what exactly should not be implemented in the base class? The methods or the service definition?

balloob commented:
I think that the services are too platform specific and therefore should be just added to the Kodi platform. Make sure to prefix them with `kodi_`.

Some other observations:

- The play_song is not necessary as one can use the play_media service.
- unset_shuffle is also not necessary, set_shuffle could just take in a parameter.
-  All three playlist services can also be folded into one `kodi_add_to_playlist` service.

alexmogavero commented:
thank you @balloob I will implement your suggestions as soon as possible

alexmogavero committed:
Services moved to kodi platform

pvizeli commented:
Ping, any updates?

alexmogavero commented:
Sorry, I have been pretty busy at work, I have started, I will probably work back on it during the weekend

alexmogavero committed:
Merge branch 'dev' into kodi

# Conflicts:
#	homeassistant/components/media_player/kodi.py

alexmogavero committed:
service play_song removed

alexmogavero committed:
removed service unset_shuffle

alexmogavero committed:
all add services merged into one

alexmogavero committed:
removed service get_artists

alexmogavero committed:
added kodi_ to service names

alexmogavero committed:
Fixed some style issues

alexmogavero commented:
I have implemented all the suggestions of @balloob , please let me know if anything is incorrect. 
In the meantime I am going to do further testing to make sure I didn't break anything.

BillyNate commented:
@alexmogavero Do you have changes stored local, or is everything in the `kodi` branch?
Maybe you could create a `kodi-wip` branch, so me or others can help you out on the requested changes (without doing duplicate work) 😉

alexmogavero committed:
Merge branch 'dev' into kodi

alexmogavero commented:
Ok I have completed all the tests. From my point of view this branch is ready to be merged. Please @balloob @pvizeli @MartinHjelmare have a look at it and let me know if anything else is needed.

balloob commented on homeassistant/components/media_player/__init__.py:
```diff
@@ -358,6 +359,11 @@ def async_service_handler(service):
             params['media_id'] = service.data.get(ATTR_MEDIA_CONTENT_ID)
             params[ATTR_MEDIA_ENQUEUE] = \
                 service.data.get(ATTR_MEDIA_ENQUEUE)
+        else:
```
Can you revert all changes to this file

balloob commented on homeassistant/components/media_player/kodi.py:
```diff
@@ -103,6 +132,38 @@ def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+        if not method:
+            return
+
+        params = {}
```
Tip, you can use:

```python
params = {key: value for key, value in service.data.items()
          if key != 'entity_id'}
```

balloob commented on homeassistant/components/media_player/kodi.py:
```diff
@@ -103,6 +132,38 @@ def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+        update_tasks = []
+        if entity.should_poll:
+            update_coro = entity.async_update_ha_state(True)
+            if hasattr(entity, 'async_update'):
```
There is no need for this check as this is always `True` for the kodi entity.

balloob commented on homeassistant/components/media_player/kodi.py:
```diff
@@ -593,6 +654,141 @@ def async_play_media(self, media_type, media_id, **kwargs):
+    @asyncio.coroutine
+    def async_set_shuffle(self, shuffle_on):
+        """Set shuffle mode, for the first player."""
+        players = yield from self._get_players()
```
There is no need to fetch the players, they are cached as `self._players` when the state is updated.

alexmogavero committed:
Removed changes in media_player __init__

alexmogavero committed:
Implemented requested changes

alexmogavero committed:
Merge branch 'dev' into kodi

alexmogavero committed:
Fixed pylint problem

alexmogavero commented:
@balloob I have implemented the changes requested. Please have a look.

balloob commented:
Looks good! :dolphin: Thanks

alexmogavero commented:
Great, thank you!

moskovskiy82 commented:
Will the docs be updates? Some sample application would be great

alexmogavero commented:
@moskovskiy82 I have added the documentation, it is now awaiting for approval. 
See home-assistant/home-assistant.github.io#2475
